### PR TITLE
Docs update & DynamicDataFrameBuilder

### PR DESCRIPTION
**Pull request by koperagen**
DynamicDataFrameBuilder here is a composition of name generator and mutable list. Useful for Kandy because they create dataframes from columns and need to know column names.

Also changed split documentation and made couple of small changes in other places, so i'd recommend to see commits one by one after https://github.com/Kotlin/dataframe/pull/434 is merged (this contains commits from that PR)

### CODE REVIEW

File: core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/api/constructors.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package org.jetbrains.kotlinx.dataframe.api
+
+import io.kotest.matchers.shouldBe
+import org.junit.Test
+
+class ConstructorsTests {
```
Comment (Jolanrensen) [inline core/src/test/kotlin/org/jetbrains/kotlinx/dataframe/api/constructors.kt:6]: Could you maybe also add a test with duplicate column names? Since DynamicDataFrameBuilder seems to be able to handle those
